config/common/cmake/chip_gn.cmake

```diff
@@ -0,0 +1,145 @@
+#
+#   Copyright (c) 2023 Project CHIP Authors
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+#
+#   @file
+#     CMake file defining 'chip' target which represents CHIP library
+#     and other optional libraries like unit tests, built with specific
+#     platform.
+#     Since CHIP doesn't provide native CMake support, ExternalProject
+#     module is used to build the required artifacts with GN meta-build
+#     system.
+#
+
+# Check or set paths
+if (NOT GN_ROOT_TARGET)
+    message(FATAL_ERROR "GN_ROOT_TARGET not defined. Please provide the path to your CHIP GN project.")
+endif()
+
+if (NOT CHIP_ROOT)
+    get_filename_component(CHIP_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../.. REALPATH)
+endif()
+
+set(CHIP_LIB_DIR ${CMAKE_CURRENT_BINARY_DIR}/lib)
+
+# Prepare CHIP libraries that the application should be linked with
+list(APPEND CHIP_LIBRARIES -lCHIP)
+
+if (CONFIG_CHIP_LIB_SHELL)
```

Comment:
I wonder if the common module could have a more clear interface. Otherwise, we end up with more difficult to understand code that splits things somewhat arbitrarily between different CMake modules and forces a user to walk through the code and detect all variable names that modify the module's behavior. Also, the common module refers directly to Kconfig variables which don't need to be defined by all CMake platforms which leads to a bit cryptic mappings:
```
if (CONFIG_CHIP_PW_RPC)
    set(CONFIG_CHIP_LIB_PW_RPC YES)
endif()
```
Also, some Kconfig variables referred to like CONFIG_CPP_STD_XXX are defined externally (in Zephyr RTOS for Zephyr-based platforms) so it may change with future Zephyr updates and require even more mappings.

Would it make sense to e.g. come up with a more abstract interface, e.g. based on macros such as:
```
matter_cflags(..)
matter_cxxflags(...)
matter_option("mdns" "platform")
...
matter_build(chip
  WITH_RPC TRUE
  WITH_SHELL TRUE
  WITH_TESTS FALSE
..
)

and let concrete platforms orchestrate the whole process and configure it based on configuration options that their platforms provide. What do you think? :)

Reply:
Yes, I understand and agree. I tried to unify the CONFIG_xx variables as much as possible. This solution required adaptation on a platform Cmake level to existing variables.
I'm not sure what should be a common part and what the platform-specific in your solution.

Reply:
I think it's a very good idea to extract the GN build and GN configuration utilities into a common module, but I think we should not assume that each CMake platform uses Kconfig and the same Kconfig naming convention. A set of CMake functions/macros that would allow me define the mapping in a single place like:
```
matter_build(chip
  RPC CONFIG_CHIP_LIB_PW_RPC
  SHELL CONFIG_CHIP_LIB_SHELL
  TESTS CONFIG_CHIP_BUILD_TESTS
)
```
seems like enough unification to me.

But I don't want to block your PR, so if you don't have time to design something like that, I can also work based on your changes someday (when I find time).

Reply:
Ok, I'll work on this.